File: typescript4/src/traversal.ts
import { Factory } from './construction';
import { Serializable } from './serialization';

export abstract class Visitor<ExpectedType extends Serializable> {
    abstract begin(obj: ExpectedType): void;
    abstract end(obj: ExpectedType): void;

    abstract owner(target: ExpectedType, ownerPropName: string): void;
    abstract verbatim<DataType>(
        getValue: (target: Serializable) => any,
        setValue: (target: Serializable, value: any) => void
    ): void;
    abstract primitive<PropType>(
        target: any, 
        propName: string, 
        fromString?: (initializer:string) => PropType): void;
    abstract scalar<ElementType extends Serializable>(
        elementFactory: Factory<ElementType>,
        propName: string
    ): void;
    abstract array<ElementType extends Serializable>(
        elementFactory: Factory<ElementType>,
        propName: string
    ): void;
    abstract map<ElementType extends Serializable>(
        elementFactory: Factory<ElementType>,
        propName: string
    ): void;
}

File: python3/elevated_objects/traversal.py
#!/usr/bin/env python3

from __future__ import annotations
import abc
import typing

from . import construction
from . import serializable

ExpectedType = typing.TypeVar('ExpectedType', bound=serializable.Serializable)

class Visitor(serializable.Visitor[ExpectedType]): 
    @abc.abstractmethod
    def begin(self, obj: ExpectedType) -> None:
        pass

    @abc.abstractmethod
    def end(self, obj: ExpectedType) -> None:
        pass

    @abc.abstractmethod
    def owner(self, target: ExpectedType, owner_prop_name: str):
        pass

    @abc.abstractmethod
    def verbatim(self,
        data_type: typing.Type, 
        target: serializable.Serializable, 
        get_value: typing.Callable [ [serializable.Serializable], typing.Any ],
        set_value: typing.Callable [ [serializable.Serializable, typing.Any ], None],
        get_prop_names: typing.Callable [ [], typing.Set[str] ]
    ) -> None:
        pass

    @abc.abstractmethod
    def primitive(self, data_type: typing.Type, target: serializable.Serializable, prop_name: str, fromString: typing.Callable[ [str], typing.Any ] = None) -> None:
        pass

    @abc.abstractmethod
    def scalar(self, element_factory: construction.Factory, target: serializable.Serializable, prop_name: str) -> None:
        pass

    @abc.abstractmethod
    def array(self, element_factory: construction.Factory, target: serializable.Serializable, prop_name: str) -> None:
        pass

    @abc.abstractmethod
    def map(self, key_type: typing.Type, element_factory: construction.Factory, target: serializable.Serializable, prop_name: str) -> None:
        pass
